File: _Bit Manipulation/cpp/190_Reverse_Bits.cpp
#include <iostream>
// Time:  O(logn) = O(32)
// Space: O(1)
using namespace std;

class Solution {
public:
    uint32_t reverseBits(uint32_t n) {
        uint32_t result = 0;
        int count = 32;
        while (count--) {
            result <<= 1;
            result |= n & 1;
            n >>= 1;
        }
        return result;
    }

    uint32_t reverseBits2(uint32_t n) {
        uint32_t result = 0;
        uint32_t mask = 1;

        for(int i=0;i<32;i++){
            result <<= 1;
            result |= n&mask;
            n>>=1;
        }
        return result;
    }
};

int main(){
    int n = 43261596;
    Solution obj;
    cout << obj.reverseBits2(n);
    return 0;
}

File: _BitManipulation/190. Reverse Bits.py
class Solution:
    # @param n, an integer
    # @return an integer
    def reverseBits2(self, n):

        sum = 0
        count = 31
        remain = n
        list = []
        while remain>0:
            if remain%2==0:
                list.append(0)
            else:
                list.append(1)
                sum = sum + 1 * 2 ** count
            remain = remain//2
            count = count -1
        return sum

    def reverseBits(self, n):
        result = 0
        mask = 1
        for i in range(32):
            result <<= 1
            result |= n&mask
            n>>=1
        return result

if __name__ == "__main__":
    n = 43261596
    obj = Solution()
    print(obj.reverseBits(n))